### PR TITLE
SI-9813 Add Java 1.6. and 1.8 java.lang.Math methods into scala.math (reopened)

### PR DESCRIPTION
**Pull request by golem131**
reopen of https://github.com/scala/scala/pull/5271

### CODE REVIEW

File: src/library/scala/math/package.scala [hunk 1]
```diff
@@ -324,4 +368,74 @@ package object math {
 
   /** @group rounding */
   def IEEEremainder(x: Double, y: Double): Double = java.lang.Math.IEEEremainder(x, y)
+
```
Comment (janekdb) [inline src/library/scala/math/package.scala:368]: exact group LGTM

File: src/library/scala/math/package.scala [hunk 2]
```diff
@@ -219,6 +222,37 @@ package object math {
   /** @group signum */
   def signum(x: Double): Double = java.lang.Math.signum(x)
 
+
+  def floorDiv(x: Int, y: Int): Int = java.lang.Math.floorDiv(x, y)
+
+
```
Comment (janekdb) [inline src/library/scala/math/package.scala]: Fix spacing

File: src/library/scala/math/package.scala [hunk 3]
```diff
@@ -219,6 +222,37 @@ package object math {
...
+  def floorMod(x: Long, y: Long): Long = java.lang.Math.floorMod(x, y)
+
+  def copySign(magnitude: Double, sign: Double): Double = java.lang.Math.copySign(magnitude, sign)
+
+  def copySign(magnitude: Float, sign: Float): Float = java.lang.Math.copySign(magnitude, sign)
+
+  def nextAfter(start: Double, direction: Double): Double = java.lang.Math.nextAfter(start, direction)
+
+  def nextAfter(start: Float, direction: Float): Float = java.lang.Math.nextAfter(start, direction)
```
Comment (janekdb) [inline src/library/scala/math/package.scala]: Why is the signature different to the Java version?

https://docs.oracle.com/javase/8/docs/api/java/lang/Math.html#nextAfter-float-double-

The Java version is `nextAfter(float start, double direction)` i.e. a deviation from the float/float or double/double pattern seen elsewhere. Find out why Java is different in this case?

File: src/library/scala/math/package.scala [hunk 4]
```diff
@@ -219,6 +222,37 @@ package object math {
...
+
+  def nextUp(d: Double): Double = java.lang.Math.nextUp(d)
+
+  def nextUp(d: Float): Float = java.lang.Math.nextUp(d)
+
+  def nextDown(d: Double): Double = java.lang.Math.nextUp(d)
```
Comment (janekdb) [inline src/library/scala/math/package.scala]: `nextDown` on RHS.

File: src/library/scala/math/package.scala [hunk 5]
```diff
@@ -219,6 +222,37 @@ package object math {
...
+
+  def nextDown(d: Float): Float = java.lang.Math.nextUp(d)
```
Comment (janekdb) [inline src/library/scala/math/package.scala]: `nextDown` on RHS.

File: src/library/scala/math/package.scala [hunk 6]
```diff
@@ -219,6 +222,37 @@ package object math {
...
+
+  def scalb(d: Double, scaleFactor: Int): Double = java.lang.Math.scalb(d, scaleFactor)
+
+
+  def scalb(d: Float, scaleFactor: Int): Float = java.lang.Math.scalb(d, scaleFactor)
```
Comment (janekdb) [inline src/library/scala/math/package.scala]: Follow the Java convention of using `f` for float parameters here and in other methods.

**Comment by janekdb** [comment]
Some comments added. I have not yet verified that all methods that could have been added have been added.

What's the plan with the method grouping? Add now or add into subsequent commit? Now is better?

**Comment by golem131** [comment]
@janekdb java.lang.Math have 73 methods and scala.math in this PR have 73 methods binded to  java.lang.Math
About grouping - what group you recommend for floorDiv/floorMod and nextUp/nextAfter?

File: src/library/scala/math/package.scala [hunk 7]
```diff
@@ -315,13 +356,83 @@ package object math {
   /** Returns the size of an ulp of the given `Double` value.
     * @group ulp
     */
-  def ulp(x: Double): Double = java.lang.Math.ulp(x)
+  def ulp(d: Double): Double = java.lang.Math.ulp(d)
```
Comment (janekdb) [inline src/library/scala/math/package.scala:359]: While `d` and `f` is now consistent with the naming convention used in the added methods which is based on the names used in Java I don't see this as an improvement because although this acts as a mnemonic to the parameter type it does not remind the user of the role of the parameter. Let the IDE and compiler remind us of the type; Use the name to describe the parameter along a different axis.

Although it's tricky to pick a name based on the role of the parameter here I would suggest `middle` based on the description of `ulp` in the Java docs. Even just reverting to `x` is better than switching to `d` and `f`.

WDYT?

Comment (janekdb) [inline src/library/scala/math/package.scala:359]: In fact this change has to be reverted or extended with a deprecated parameter name to be binary compatible. To maintain the focus of this PR to the addition of missing methods defer this renaming to a separate PR.

File: src/library/scala/math/package.scala [hunk 8]
```diff
@@ -215,9 +218,37 @@ package object math {
...
+
+  def copySign(magnitude: Double, sign: Double): Double = java.lang.Math.copySign(magnitude, sign)
+
+  def copySign(magnitude: Float, sign: Float): Float = java.lang.Math.copySign(magnitude, sign)
+
+  def nextAfter(start: Double, direction: Double): Double = java.lang.Math.nextAfter(start, direction)
+
+  def nextAfter(start: Float, direction: Double): Float = java.lang.Math.nextAfter(start, direction)
+
+  def nextUp(d: Double): Double = java.lang.Math.nextUp(d)
```
Comment (janekdb) [inline src/library/scala/math/package.scala:241]: I'd like to see if we can take this opportunity to add some value to the parameter name. As mentioned below `d` and `f` provide no additional information that the typing doesn't already provide. Can we have more meaningful names?

Happy to make some suggestions later.

File: src/library/scala/math/package.scala (hunk 7 above)
Comment (golem131) [inline src/library/scala/math/package.scala:359]: I agree with you, and at first in Java docs you can see - https://docs.oracle.com/javase/8/docs/api/java/lang/Math.html#ulp-double-

ulp(double d)/ulp(float f)

and second, i think, if we use convention, that we base naming on Java docs/Java source code, we need to do this in all places in file, not just in new bindings.

File: src/library/scala/math/package.scala [hunk 9]
```diff
@@ -215,9 +218,37 @@ package object math {
     */
   def signum(x: Long): Long     = java.lang.Long.signum(x)
   /** @group signum */
-  def signum(x: Float): Float   = java.lang.Math.signum(x)
+  def signum(f: Float): Float   = java.lang.Math.signum(f)
```
Comment (janekdb) [inline src/library/scala/math/package.scala:221]: The parameter name cannot be changed with a deprecation annotation. This is out of scope for this PR don't you think?

Revert to `x`.

File: src/library/scala/math/package.scala [hunk 10]
```diff
@@ -215,9 +218,37 @@ package object math {
...
   /** @group signum */
-  def signum(x: Double): Double = java.lang.Math.signum(x)
+  def signum(d: Double): Double = java.lang.Math.signum(d)
```
Comment (janekdb) [inline src/library/scala/math/package.scala:223]: Would require deprecation annotation.

Revert to `x`.

**Comment by janekdb** [comment]
Confirmed all methods present in the Java 8 class are present here. :+1: 

Changes to debate further,
1. Follow the established pattern of using `x` instead of `d` and `f` for parameter names

File: src/library/scala/math/package.scala (hunk 8 above)
Comment (janekdb) [inline src/library/scala/math/package.scala:241]: I think `x` is good enough.

File: src/library/scala/math/package.scala (hunk 7 above)
Comment (janekdb) [inline src/library/scala/math/package.scala:359]: So I think we are in agreement: Use `x`.

File: src/library/scala/math/package.scala [hunk 11]
```diff
@@ -315,13 +356,83 @@ package object math {
...
+
+  /**
+    * @group exact
+    */
+  def multiplyExact(x: Long, y: Long): Long = java.lang.Math.multiplyExact(x, y)
+
+  /**
+    * @group exact
+    */
+  def incrementExact(a: Int): Int = java.lang.Math.incrementExact(a)
```
Comment (janekdb) [inline src/library/scala/math/package.scala:406]: `a` is carried over from the Java 8 names. Maybe the Java engineers are working to some hidden convention but I can't figure it out. For example in Java

```
log(double a)
log10(double a)
log1p(double x)
toIntExact(long value)
```

Possibly there are some conventions from the world of floating point arithmetic at play here.

Overall I would just stick with `x` unless there is reason not too.

File: src/library/scala/math/package.scala [hunk 12]
```diff
@@ -215,9 +218,37 @@ package object math {
     */
   def signum(x: Long): Long     = java.lang.Long.signum(x)
   /** @group signum */
-  def signum(x: Float): Float   = java.lang.Math.signum(x)
+  def signum(f: Float): Float   = java.lang.Math.signum(f)
   /** @group signum */
-  def signum(x: Double): Double = java.lang.Math.signum(x)
+  def signum(d: Double): Double = java.lang.Math.signum(d)
+
+  def floorDiv(x: Int, y: Int): Int = java.lang.Math.floorDiv(x, y)
```
Comment (janekdb) [inline src/library/scala/math/package.scala:225]: @golem131 - WDYT of _Modulus and Quotient_ as the group name for `floorDiv` and `floorMod`? This fits in with the existing group names and is clear.

While this group contains just these four methods a short note in `@groupdesc` could be added

> Calculate quotient values by rounding to negative infinity.

File: src/library/scala/math/package.scala [hunk 13]
```diff
@@ -215,9 +218,37 @@ package object math {
...
+
+  def floorMod(x: Int, y: Int): Int = java.lang.Math.floorMod(x, y)
+
+  def floorMod(x: Long, y: Long): Long = java.lang.Math.floorMod(x, y)
+
+  def copySign(magnitude: Double, sign: Double): Double = java.lang.Math.copySign(magnitude, sign)
+
+  def copySign(magnitude: Float, sign: Float): Float = java.lang.Math.copySign(magnitude, sign)
+
+  def nextAfter(start: Double, direction: Double): Double = java.lang.Math.nextAfter(start, direction)
```
Comment (janekdb) [inline src/library/scala/math/package.scala:237]: @golem131 - Suggestion for `next*` group name: _Adjacent Floats_

It's short and to the point. For anyone hunting out the functionality of a `next*` method this is a strong cue.

File: src/library/scala/math/package.scala [hunk 14]
```diff
@@ -215,9 +218,37 @@ package object math {
...
+
+  def floorDiv(x: Int, y: Int): Int = java.lang.Math.floorDiv(x, y)
+
+  def floorDiv(x: Long, y: Long): Long = java.lang.Math.floorDiv(x, y)
+
+  def floorMod(x: Int, y: Int): Int = java.lang.Math.floorMod(x, y)
+
+  def floorMod(x: Long, y: Long): Long = java.lang.Math.floorMod(x, y)
+
+  def copySign(magnitude: Double, sign: Double): Double = java.lang.Math.copySign(magnitude, sign)
```
Comment (janekdb) [inline src/library/scala/math/package.scala:233]: @golem131 `copySign` can be included in the "Signs" group. The group name is `signum` so rename that to `@group signs` to make it less specific to `signum`.

File: src/library/scala/math/package.scala [hunk 15]
```diff
@@ -215,9 +218,37 @@ package object math {
...
+
+  def nextUp(d: Double): Double = java.lang.Math.nextUp(d)
+
+  def nextUp(f: Float): Float = java.lang.Math.nextUp(f)
+
+  def nextDown(d: Double): Double = java.lang.Math.nextDown(d)
+
+  def nextDown(f: Float): Float = java.lang.Math.nextDown(f)
+
+  def scalb(d: Double, scaleFactor: Int): Double = java.lang.Math.scalb(d, scaleFactor)
```
Comment (janekdb) [inline src/library/scala/math/package.scala:249]: @SethTisue  - Add a new group called `scaling` for `scalb`

```
@group scaling
```

Use this description,

```
@groupdesc scaling Scaling with rounding guarantees
```

**Comment by janekdb** [comment]
@golem131 - Comments on group names added.

**Comment by janekdb** [comment]
@golem131 - The final commit message should follow the commit message guidelines found here: 

```
https://github.com/scala/scala/blob/2.12.x/CONTRIBUTING.md
Clean commits, clean history
```

File: src/library/scala/math/package.scala [hunk 16]
```diff
@@ -63,6 +63,9 @@ package scala
   *
   * @groupname randomisation Pseudo Random Number Generation
   * @groupprio randomisation 130
+  *
+  * @groupname exact Functions to avoid overflow on primitive types
```
Comment (janekdb) [inline src/library/scala/math/package.scala:67]: @SethTisue - Change to

```
@groupname exact Exact Arithmetic
@groupdesc exact Integral addition, multiplication, stepping and conversion throwing ArithmeticException instead of underflowing or overflowing
```

**Comment by janekdb** [comment]
@SethTisue - Reviewed. Changes suggested.

**Comment by janekdb** [comment]
@SethTisue - Grant me push permission for this PR and I will happily take it over the finishing line.

**Comment by SethTisue** [comment]
@janekdb do I have that power? is there a GitHub feature I don't know about? if not, just submit your own PR that includes @golem131's commits

**Comment by janekdb** [comment]
Noone has that power as far as I can tell! Progressing with #5382

**Comment by SethTisue** [comment]
@janekdb well hey looky here https://github.com/blog/2247-improving-collaboration-with-forks . too late for this PR though ("pull requests created before the option was available will default to not allowing collaboration"), so let's stick with the new one

**Comment by janekdb** [comment]
@SethTisue - Well, well, well! Nice find.

File: src/library/scala/math/package.scala (hunk 8 above)
Comment (janekdb) [reply src/library/scala/math/package.scala:241]: Done in #5382

File: src/library/scala/math/package.scala (hunk 9 above)
Comment (janekdb) [reply src/library/scala/math/package.scala:221]: Done in #5382

File: src/library/scala/math/package.scala (hunk 10 above)
Comment (janekdb) [reply src/library/scala/math/package.scala:223]: Done in #5382

File: src/library/scala/math/package.scala (hunk 11 above)
Comment (janekdb) [reply src/library/scala/math/package.scala:406]: Done in #5382

File: src/library/scala/math/package.scala (hunk 12 above)
Comment (janekdb) [reply src/library/scala/math/package.scala:225]: Done in #5382

File: src/library/scala/math/package.scala (hunk 13 above)
Comment (janekdb) [reply src/library/scala/math/package.scala:237]: Done in #5382

File: src/library/scala/math/package.scala (hunk 14 above)
Comment (janekdb) [reply src/library/scala/math/package.scala:233]: Done in #5382

File: src/library/scala/math/package.scala (hunk 15 above)
Comment (janekdb) [reply src/library/scala/math/package.scala:249]: Done in #5382

File: src/library/scala/math/package.scala (hunk 16 above)
Comment (janekdb) [reply src/library/scala/math/package.scala:67]: Done in #5382
